staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder.go

```diff
@@ -0,0 +1,408 @@
+		return spec.MustCreateRef(definitionPrefix + common.EscapeJsonPointer(defName))
+	})
+}
+
+// addTypeMetaProperties adds Kubernetes-specific type meta properties to input schema:
+//     apiVersion and kind
+func addTypeMetaProperties(s *spec.Schema) {
+	s.SetProperty("apiVersion", getDefinition(typeMetaType).SchemaProps.Properties["apiVersion"])
+	s.SetProperty("kind", getDefinition(typeMetaType).SchemaProps.Properties["kind"])
+}
+
+// buildListSchema builds the list kind schema for the CRD
+func (b *builder) buildListSchema() *spec.Schema {
+	name := definitionPrefix + openapi.FriendlyName(fmt.Sprintf("%s/%s/%s", b.group, b.version, b.kind))
+	doc := fmt.Sprintf("List of %s. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md", b.plural)
+	s := new(spec.Schema).WithDescription(fmt.Sprintf("%s is a list of %s", b.listKind, b.kind)).
+		WithRequired("items").
+		SetProperty("items", *spec.ArrayProperty(spec.RefSchema(name)).WithDescription(doc)).
+		SetProperty("metadata", getDefinition(listMetaType))
+	addTypeMetaProperties(s)
+	s.AddExtension(endpoints.ROUTE_META_GVK, []map[string]string{
+		{
+			"group":   b.group,
+			"version": b.version,
+			"kind":    b.listKind,
+		},
+	})
+	return s
+}
+
+// getOpenAPIConfig builds config which wires up generated definitions for kube-openapi to consume
+func (b *builder) getOpenAPIConfig() *common.Config {
+	return &common.Config{
+		ProtocolList: []string{"https"},
+		Info: &spec.Info{
+			InfoProps: spec.InfoProps{
+				Title:   "Kubernetes CRD Swagger",
+				Version: "v0.1.0",
+			},
+		},
+		CommonResponses: map[int]spec.Response{
+			401: {
+				ResponseProps: spec.ResponseProps{
+					Description: "Unauthorized",
+				},
+			},
+		},
+		GetOperationIDAndTags: openapi.GetOperationIDAndTags,
+		GetDefinitionName: func(name string) (string, spec.Extensions) {
+			buildDefinitions.Do(buildDefinitionsFunc)
+			return namer.GetDefinitionName(name)
+		},
+		GetDefinitions: func(ref common.ReferenceCallback) map[string]common.OpenAPIDefinition {
+			def := generatedopenapi.GetOpenAPIDefinitions(ref)
+			def[fmt.Sprintf("%s/%s.%s", b.group, b.version, b.kind)] = common.OpenAPIDefinition{
+				Schema: *b.schema,
+			}
+			def[fmt.Sprintf("%s/%s.%s", b.group, b.version, b.listKind)] = common.OpenAPIDefinition{
+				Schema: *b.listSchema,
+			}
+			return def
+		},
+	}
+}
+
+func newBuilder(crd *apiextensions.CustomResourceDefinition, version string, schema *spec.Schema) *builder {
+	b := &builder{
+		schema:     &spec.Schema{},
+		listSchema: &spec.Schema{},
+		ws:         &restful.WebService{},
+
+		group:    crd.Spec.Group,
+		version:  version,
+		kind:     crd.Spec.Names.Kind,
+		listKind: crd.Spec.Names.ListKind,
+		plural:   crd.Spec.Names.Plural,
+	}
+	if crd.Spec.Scope == apiextensions.NamespaceScoped {
+		b.namespaced = true
+	}
+	// Pre-build schema with Kubernetes native properties
+	if schema != nil {
+		b.schema = b.buildKubeNative(schema)
```

Comment:
this is odd. Instead we should use an empty `{"type":"object"}` spec as basis instead of nil. But we still have to add the metadata and TypeMeta fields.